### PR TITLE
external QEMU snap support

### PR DESCRIPTION
**Pull request by mihalicyn**
requires https://github.com/canonical/lxd/pull/13742

https://warthogs.atlassian.net/browse/LXD-1164

### CODE REVIEW

File: lxd-qemu-snap/README.md
Comment (simondeziel) [inline lxd-qemu-snap/README.md:42]: Aren't those 2 rules already covered by the `/dev/dri/{,**} rw,` one before?

File: lxd-qemu-snap/README.md
Comment (simondeziel) [inline lxd-qemu-snap/README.md]:
```suggestion
# lspci | grep VGA shows:
```

File: snapcraft/commands/daemon.start
Comment (simondeziel) [inline snapcraft/commands/daemon.start]: I see pulseaudio in there and I wonder about pipewire?

Comment (mihalicyn) [reply snapcraft/commands/daemon.start]: it's because QEMU was failing on attempt to load necessary libs from `${SNAP_CURRENT}/opt/external-qemu/lib/${ARCH}/pulseaudio`. Do you think we should add `${SNAP_CURRENT}/opt/external-qemu/lib/x86_64-linux-gnu/pipewire-0.3/` too?

File: lxd-qemu-snap/README.md
Comment (mihalicyn) [reply lxd-qemu-snap/README.md]: it's not a mistake, actually, I wanted to show what's behind `67:00` from `ls -la /dev/dri/by-path/*-render`.

Let's do it even better:
`lspci | grep -E "(3D|VGA)"`

What do you think?

File: lxd-qemu-snap/README.md
Comment (mihalicyn) [reply lxd-qemu-snap/README.md:42]: sure, let's do even better:
```
    /snap/lxd/*/gpu-2404/** mr,
    /dev/dri/ r,
    /dev/dri/card[0-9]* rw,
    /dev/dri/renderD[0-9]* rw,
    /run/udev/data/c226:[0-9]* r,  # 226 drm
    /sys/devices/** r,
    /sys/bus/** r,
```

File: snapcraft/commands/daemon.start
Comment (simondeziel) [reply snapcraft/commands/daemon.start]: I don't know pipewire at all, but maybe? If so, could we use a shell glob to avoid hardcoding the version in the dir name?

File: lxd-qemu-snap/README.md
Comment (simondeziel) [reply lxd-qemu-snap/README.md]: Oh, I see, I hadn't read the previous line properly. But sure, that `grep -E` LGTM.

File: lxd-qemu-snap/README.md
Comment (simondeziel) [inline lxd-qemu-snap/README.md]:
```suggestion
rm -f qemu-for-lxd_*.snap
```
For those blindly copy-n-pasting code snippets ;)

**Review by simondeziel — COMMENTED** [review]
LGTM in general even if I don't like that `/opt` I won't object to it ;)

File: snapcraft/commands/daemon.start
Comment (mihalicyn) [reply snapcraft/commands/daemon.start]: >I don't know pipewire at all, but maybe?

Unfortunately, I don't know too ;-)

>If so, could we use a shell glob to avoid hardcoding the version in the dir name?

Hm, something like:
```
LD_LPATH_PIPEWIRE="$(readlink -f "${SNAP_CURRENT}"/opt/external-qemu/lib/"${ARCH}"/pipewire-*/)"
export LD_LIBRARY_PATH="${SNAP_CURRENT}/opt/external-qemu/lib/${ARCH}:${SNAP_CURRENT}/opt/external-qemu/lib/${ARCH}/pulseaudio:${SNAP_CURRENT}/opt/external-qemu/lib/${ARCH}/ceph:${LD_LPATH_PIPEWIRE:+${LD_LPATH_PIPEWIRE}:}${LD_LIBRARY_PATH}"
```

?

Comment (simondeziel) [reply snapcraft/commands/daemon.start]: Yeah, that `readlink` works.

File: lxd-qemu-snap/README.md
Comment (mihalicyn) [reply lxd-qemu-snap/README.md]: Fixed. Thanks!

File: snapcraft/commands/daemon.start
Comment (mihalicyn) [reply snapcraft/commands/daemon.start]: Have fixed!

**Comment by mihalicyn** [comment]
Reworked to use connect/disconnect hooks.

**Review by simondeziel — DISMISSED** [review]
LGTM, thanks!

**Comment by mihalicyn** [comment]
Have added reload logic to connect/disconnect hooks.

File: lxd-qemu-snap/snapcraft.yaml
Comment (tomponline) [inline lxd-qemu-snap/snapcraft.yaml:34]: Is this section needed? What is it used for?

Comment (mihalicyn) [reply lxd-qemu-snap/snapcraft.yaml:34]: It does a cleanup of libraries on the prime step in snap to prevent collisions with what `gpu-2404` snap provides.

File: snapcraft.yaml
Comment (tomponline) [inline snapcraft.yaml:1519]: Does this not get provided by the connected gpu-2404 snap? Whyy do we need this pulled here too?

Comment (mihalicyn) [reply snapcraft.yaml:1519]: no, this is not provided by `gpu-2404`. This repo (https://github.com/canonical/gpu-snap) contains tools to help integration with `gpu-2404` interface snaps. This step, specifically, needed to clean up libraries on the prime step to prevent duplicates.

Comment (mihalicyn) [reply snapcraft.yaml:1519]: You can read about it in the official guide https://mir-server.io/docs/the-gpu-2404-snap-interface

File: lxd-qemu-snap/snapcraft.yaml
Comment (simondeziel) [inline lxd-qemu-snap/snapcraft.yaml]: Should the `-cleanup` script also be primed?
The [gpu-2404-cleanup](https://github.com/canonical/gpu-snap/blob/main/bin/gpu-2404-cleanup) script is a symlink to [graphics-core22-cleanup](https://github.com/canonical/gpu-snap/blob/main/bin/graphics-core22-cleanup). I don't know if the `prime` logic is smart enough to cope with that.

File: snapcraft.yaml
Comment (simondeziel) [inline snapcraft.yaml]: Is the `-cleanup` needed here?

File: lxd-qemu-snap/snapcraft.yaml
Comment (mihalicyn) [reply lxd-qemu-snap/snapcraft.yaml]: I don't think that we need to prime it as we just want to use it once during snap build time. And we doing it in the line `${CRAFT_PART_SRC}/bin/gpu-2404-cleanup mesa-2404 nvidia-2404`

Comment (simondeziel) [reply lxd-qemu-snap/snapcraft.yaml]: OK, I hadn't realize it was a copy-n-paste from https://mir-server.io/docs/the-gpu-2404-snap-interface
Have you been able to test this? Cause when reading the `gpu-2404-cleanup` script, I think it will error out due to `nvidia-24.04` not existing in the [lists directory](https://github.com/canonical/gpu-snap/tree/main/lists).

Comment (mihalicyn) [reply lxd-qemu-snap/snapcraft.yaml]: nope, I'm not specifically tested how this script works. I haven't seen any build time errors or warnings.

Comment (mihalicyn) [reply lxd-qemu-snap/snapcraft.yaml]: I believe that `nvidia-2404` will be added soon and as we pull it in dynamically (without tag/version specifier) it will work properly.

**Review by simondeziel — DISMISSED** [review]

File: snapcraft.yaml
Comment (tomponline) [inline snapcraft.yaml]: Looking at the other plugs, and the snap feature options on https://snapcraft.io/lxd it seems like we usually use `noun.adjective` format, so this would be more in keeping to be called `qemu-external` in my view.

**Review by tomponline — CHANGES_REQUESTED** [review]
Apart from the nit around naming consistency this looks good, thanks!

File: snapcraft.yaml
Comment (mihalicyn) [reply snapcraft.yaml]: Thanks! Have changed that.

File: lxd-qemu-snap/snapcraft.yaml
Comment (simondeziel) [inline lxd-qemu-snap/snapcraft.yaml:48]: Would doing this via the [`organize:`](https://snapcraft.io/docs/snapcraft-parts-metadata#organize) section work?

**Review by tomponline — APPROVED** [review]
Thanks!

**Review by simondeziel — APPROVED** [review]
Wasn't fast enough to provide feedback before the merge, but maybe check if `organize:` would work?

File: lxd-qemu-snap/snapcraft.yaml
Comment (mihalicyn) [reply lxd-qemu-snap/snapcraft.yaml:48]: Can we create a symlink using the `organize` step? I haven't known about that.

Comment (simondeziel) [reply lxd-qemu-snap/snapcraft.yaml:48]: No, it seems to only do renames.

Comment (mihalicyn) [reply lxd-qemu-snap/snapcraft.yaml:48]: ah, then I think the `organize` thing is not very helpful in this case. We don't want to have `usr/lib` at the end. It's just a temporary hack to provide a cleanup script with what it expects to see. At the end we remove `usr/lib` symlink and all libraries live in `lib`.

**Pull request merged**
